### ISSUE TITLE
Cineon cleanup

### ISSUE THREAD

**Issue by jeremyselan** [OpenImageIO/oiio#154]
The cineon and dpx codepaths share many common chunks of code. However, the dpx has had many bugfixes that were not ported to cineon.

For example, if we compare cineon.imageio/libcineon/BaseTypeConverter.h dpx.imageio/libdpx/BaseTypeConverter.h  we see the dpx has code to preserve white point scaling.  However, cineon does not have that code (and it should!).

Additionally, cineon has a compile warning that should be addressed:

`In file included from /net/homedirs/jeremys/git/oiio.js/src/cineon.imageio/libcineon/Writer.cpp:41:                                                                   
/net/homedirs/jeremys/git/oiio.js/src/cineon.imageio/libcineon/WriterInternal.h: In function ‘void cineon::WritePackedMethod(IB*, IB*, int, bool, cineon::BufferAccess&) [with IB = unsigned int, int BITDEPTH = 32]’:                                                                                                                      
/net/homedirs/jeremys/git/oiio.js/src/cineon.imageio/libcineon/WriterInternal.h:273:   instantiated from ‘int cineon::WriteBuffer(cineon::OutStream*, cineon::DataSize, void*, cineon::U32, cineon::U32, int, cineon::Packing, bool, int, char*, bool&) [with IB = unsigned int, int BITDEPTH = 32, bool SAMEBUFTYPE = true]’               
/net/homedirs/jeremys/git/oiio.js/src/cineon.imageio/libcineon/Writer.cpp:286:   instantiated from here                                                               
/net/homedirs/jeremys/git/oiio.js/src/cineon.imageio/libcineon/WriterInternal.h:145: warning: right shift count is negative                                           
/net/homedirs/jeremys/git/oiio.js/src/cineon.imageio/libcineon/WriterInternal.h: In function ‘void cineon::WritePackedMethod(IB*, IB*, int, bool, cineon::BufferAccess&) [with IB = long unsigned int, int BITDEPTH = 64]’:                                                                                                                 
/net/homedirs/jeremys/git/oiio.js/src/cineon.imageio/libcineon/WriterInternal.h:273:   instantiated from ‘int cineon::WriteBuffer(cineon::OutStream*, cineon::DataSize, void*, cineon::U32, cineon::U32, int, cineon::Packing, bool, int, char*, bool&) [with IB = long unsigned int, int BITDEPTH = 64, bool SAMEBUFTYPE = true]’          
/net/homedirs/jeremys/git/oiio.js/src/cineon.imageio/libcineon/Writer.cpp:293:   instantiated from here                                                               
/net/homedirs/jeremys/git/oiio.js/src/cineon.imageio/libcineon/WriterInternal.h:145: warning: right shift count is negative
`

I am marking this as a bug, due to the issue that at least one of these changes (the bit depth 10->16 promotion) will affect image results.

**Comment by devernay** [comment]
another issue with Cineon is that standard Cineon files such as the Kodak Digital LAD http://motion.kodak.com/motion/support/technical_information/lab_tools_and_techniques/digital_lad/default.htm ar read as:
```
    oiio:ColorSpace: "GammaCorrected"
    oiio:Gamma: 0
```
The following condition, which appears twice in cineoninput.cpp:
```
if (!isinf (m_cin.header.Gamma ()))
```
should probably be:
```
if (!isinf (m_cin.header.Gamma ()) && m_cin.header.Gamma () != 0.)
```

**Comment by lgritz** [comment]
@devernay , see #1466 for the minor fix you pointed out with the gamma.

Did you just append here because it was related to Cineon, or are you having trouble corresponding to the original reason for this issue? Because it's a rather old ticket and nobody has complained about it in years (Jeremy, who opened this issue, worked for my employers, so if I'm not getting complaints internally, Jeremy's original concerns are essentially theoretical, I don't think we use Cineon extensively enough to spend time on it). I'd be happy for somebody to fix, but without any serious complaints I can't justify spending the time on it myself.

**Comment by devernay** [comment]
I append here because I always search for old open issues before opening a new one, but finally thought it deserved its own issue (#1463), because after this first Cineon failure (which happened with Kodak DLAD), I stumbled on more failures with Nuke-create cineon files from the Nuke tutorial assets. I'm still not sure about what "gamma" means in Cineon files, which seem to always be log-eencoded. Is this gamma supposed to replace log encoding, or to be applied afterwards?

**Comment by malcolmhumphreys** [comment]
From the DPX spec it suggests that Gamma is for a undefined video signal (ie not Log) and requires other header values.

> 1) For the undefined video signal standard, it is necessary to specify the following fields that would otherwise be fully specified by selecting one of the video signal standards:
68 Gamma
69 Black level code value
70 Black gain
71 Breakpoint
72 Reference white level code value

> 4.38 gamma: Defines the power law expo- nent that represents the gamma correction applied to a video image. In the expression Y = X1/gamma, the default gamma for NTSC is 2.2. [68]

From how I read the DPX spec these headers values are mutually exclusive and don't get applied together. Also these header fields are apart of the 'Television information' industry specific header section.

**Comment by malcolmhumphreys** [comment]
This is what it says in the DPX spec, I would suggest that for CIN that the headers by in large behave the same.

**Comment by lgritz** [comment]
Well, then as a practical matter, are all Cineon files in the real world encoded as log, and so our cineon reader should just always declare the color space to be "KodakLog" and forget about the gamma value?

**Comment by devernay** [comment]
It sounds reasonable. That's what Nuke, ImageMagick and GraphicMagick do, e.g.:
http://hg.code.sf.net/p/graphicsmagick/code/file/7fa54cb64ff8/coders/cineon.c#l1141

According to GraphicsMagick, the gamma is this of the *capture device*, not of the encoded values, which makes sense since it's in the "Origination Info" header, and not in the "Image Info" header.

Another important information: it seems like the header data is bigendian, and thus all 32-bit values have to be swabbed on x86, but the libcineon source code seems to only swabs the image data, not the header data (thus the crazy values read from nuke-encoded cineons). Did I miss something?

GraphicsMagick swabs the header members too:
http://hg.code.sf.net/p/graphicsmagick/code/file/7fa54cb64ff8/coders/cineon.c#l512

ImageMagick also swabs header members (it sets `image->endian`, and then uses `ReadBlob*()` which swabs):
https://github.com/ImageMagick/ImageMagick/blob/master/coders/cin.c#L450

**Comment by lgritz** [comment]
I think this was fixed in #1466, but this issue didn't auto-close.
